src/java.desktop/share/classes/javax/swing/plaf/synth/SynthGraphicsUtils.java

```diff
@@ -731,4 +744,47 @@ public int getIconHeight() {
             return synthIcon.getIconHeight(context);
         }
     }
+
+    /*
+     * Repaints all the components with the mnemonics in the given window and all its owned windows.
+     */
+    static void repaintMnemonicsInWindow(final Window w) {
+        if (w == null || !w.isShowing()) {
+            return;
+        }
+
+        final Window[] ownedWindows = w.getOwnedWindows();
+        for (final Window element : ownedWindows) {
+            repaintMnemonicsInWindow(element);
+        }
+
+        repaintMnemonicsInContainer(w);
+    }
+
+    /*
+     * Repaints all the components with the mnemonics in container.
+     * Recursively searches for all the subcomponents.
+     */
+    static void repaintMnemonicsInContainer(final Container cont) {
+        final Component[] elements = cont.getComponents();
+        for (final Component c : elements) {
+            if (c == null || !c.isVisible()) {
+                continue;
+            }
+
+            if (c instanceof AbstractButton && ((AbstractButton) c).getMnemonic() != '\0') {
+                c.repaint();
+                continue;
+            }
+
+            if (c instanceof JLabel && ((JLabel) c).getDisplayedMnemonic() != '\0') {
+                c.repaint();
+                continue;
+            }
+
+            if (c instanceof Container) {
```

Comment:
JComponent extends Container ..  so this will traverse everything in the Swing UI.
Is there any pattern elsewhere in Swing you can use to short-circuit this ?

Reply:
since this method should be recursing through all of a component's subcomponents, shouldn't there not even be an if check here?

Reply:
> JComponent extends Container .. so this will traverse everything in the Swing UI.
Is there any pattern elsewhere in Swing you can use to short-circuit this ?

Couldn't recall at this moment.

Reply:
> since this method should be recursing through all of a component's subcomponents, shouldn't there not even be an if check here?

If I understand correctly, you mean to say we don't require an if condition here. But the if condition is to identify the components in container which needs to be repainted for mnemonic related changes. 
So if a component is an instance of container, all the sub-components are fetched and if any of sub-component is either AbstractButton or JLabel instance, it will be repainted based on mnemonic condition otherwise will be ignored (if not a container itself).